gpt4all-chat/database.cpp

```diff
@@ -1038,6 +1038,53 @@ void Database::handleDocumentError(const QString &errorMessage,
     qWarning() << errorMessage << document_id << document_path << error;
 }
 
+static QString recursiveTextSplitter(QString &text, int &splitLevel, size_t maxChars,
+    const QList<QString> &splitters, int splitterIndex = 0)
+{
+    // If our text is below the limit, then return it
+    if (text.size() <= maxChars) {
+        splitLevel = splitterIndex;
+        const QString chunk = text;
+        text.clear();
+        return chunk;
+    }
+
+    // If we can't split further, then slice and return what we can
+    if (splitterIndex >= splitters.size()) {
+        splitLevel = -1;
+        const QString chunk = text.first(maxChars);
+        text.remove(0, chunk.size());
+        return chunk;
+    }
```

Comment:
I would prefer if we did this before recursive calls instead of inside of the call so we don't silently ignore the case where the original caller passes no "splitters."

Reply:
We still have to do it inside as it is possible that we'll be asked to chunk a very long string that has no splitters. The -1 indicates that splitting was not possible in such a case.